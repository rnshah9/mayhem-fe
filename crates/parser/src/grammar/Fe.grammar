file_input: NEWLINE ENDMARKER | module_stmt+ ENDMARKER

module_stmt: import_stmt | type_def | contract_def | func_def

########################### path ######################################

path: NAME ('::' NAME)*

trailing_delim_path: path '::'

########################### use_stmt ##################################

use_stmt: 'use' use_tree NEWLINE

use_tree: (simple_use_tree | glob_use_tree | nested_use_tree)

simple_use_tree: path ['as' NAME]
glob_use_tree: trailing_delim_path '*'
nested_use_tree: trailing_delim_path '{' [use_tree (',' use_tree)* [',']] '}'

########################### type_def #################################

type_def: 'type' NAME '=' type_desc NEWLINE

########################### contract_def #############################

contract_def: 'contract' NAME '{' contract_stmt* '}'
contract_stmt: contract_field | event_def | func_def
contract_field: [contract_field_qual] NAME ':' type_desc NEWLINE

event_def: 'event' NAME '{' event_field* '}'
event_field: [event_field_qual] NAME ':' type_desc NEWLINE

func_def:
    [func_qual] 'fn' NAME '(' [arg_list] ')' ['->' base_type] ':' block
arg_list: arg_def (',' arg_def)* [',']
arg_def: NAME ':' type_desc | 'self'

type_desc: map_type | base_type
map_type:
    # This is required because of the ambiguous right-shift token
    'map' '<' base_type ',' 'map' '<' base_type ',' type_desc '>>' |
    'map' '<' base_type ',' type_desc '>'
base_type: NAME arr_list
arr_list: ('[' NUMBER ']')*

contract_field_qual: 'const' | 'pub'
event_field_qual: 'idx'
func_qual: 'pub'

########################### func_stmt ###############################

func_stmt:  compound_stmt | simple_stmt
simple_stmt: small_stmt (';' small_stmt)* [';']

small_stmt: return_stmt | assert_stmt | emit_stmt | break_stmt |
            continue_stmt | revert_stmt | vardecl_stmt | assign_stmt | augassign_stmt |
            exprs

return_stmt: 'return' [exprs]
assert_stmt: 'assert' expr [',' expr]
emit_stmt: 'emit' expr

break_stmt: 'break'
continue_stmt: 'continue'
revert_stmt: 'revert'

vardecl_stmt: target ':' type_desc ['=' expr]
assign_stmt: (targets '=')+ exprs
augassign_stmt: target augassign exprs

augassign: '+=' | '-=' | '*=' | '/=' | '%=' | '&=' | '|=' | '^='
           | '<<=' | '>>=' | '**=' | '//='

compound_stmt: if_stmt | while_stmt | for_stmt

if_stmt: 'if' expr block [else_block]
else_block: 'else'  block

while_stmt: 'while' expr block [else_block]

for_stmt: 'for' targets 'in' exprs block [else_block]

block: '{' (simple_stmt | func_stmt+) '}'

exprs: expr (',' expr)* [',']

expr: disjunct ['if' disjunct 'else' expr]
disjunct: conjunct ('or' conjunct)*
conjunct: comparison ('and' comparison)*
comparison: 'not'* bitwise_or (comp_op bitwise_or)*
comp_op: '<' | '<=' | '==' | '>=' | '>' | '!=' | ['not'] 'in' | 'is' ['not']

bitwise_or: bitwise_xor ('|' bitwise_xor)*
bitwise_xor: bitwise_and ('^' bitwise_and)*
bitwise_and: shift_expr ('&' shift_expr)*
shift_expr: sum (('<<'|'>>') sum)*

sum: term (('+' term | '-' term))*
term: factor (('*' factor | '/' factor | '//' factor | '%' factor))*
factor: ('+' | '-' | '~') factor | power
power: primary '**' factor | primary
primary: atom (attr_tail | index_tail | call_tail)*

slices: slice (',' slice)* [',']
slice: [expr] ':' [expr] [':' [expr]] | expr

atom: list | listcomp | group | tuple | NAME | STRING+ | NUMBER | '...'
list: '[' [exprs] ']'
listcomp: '[' expr for_if_clauses ']'
tuple: '(' [exprs] ')'
group: '(' expr ')'
for_if_clauses: ('for' targets 'in' expr ('if' expr)*)+

args: kwargs | expr [',' args]
kwargs: kwarg (',' kwarg)*
kwarg: NAME '=' expr

targets: target (',' target)* [',']
target: atom t_tail+ | t_atom t_tail*
t_atom: NAME | '(' [targets] ')' | '[' [targets] ']'

t_tail: call_tail* (attr_tail | index_tail)

attr_tail: '.' NAME
index_tail: '[' slices ']'
call_tail: '(' [args] ')'
