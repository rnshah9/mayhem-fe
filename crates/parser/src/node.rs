pub use fe_common::{Span, Spanned};
use serde::{Deserialize, Serialize};
use std::sync::atomic::{AtomicU32, Ordering};

#[derive(Debug, PartialEq, Copy, Clone, Hash, Eq, Default, PartialOrd, Ord)]
pub struct NodeId(u32);

impl NodeId {
    pub fn create() -> Self {
        static NEXT_ID: AtomicU32 = AtomicU32::new(0);
        Self(NEXT_ID.fetch_add(1, Ordering::Relaxed))
    }

    pub fn dummy() -> Self {
        Self(u32::MAX)
    }

    pub fn is_dummy(self) -> bool {
        self == Self::dummy()
    }
}

#[derive(Serialize, Deserialize, Debug, PartialEq, Eq, Hash, Clone)]
pub struct Node<T> {
    pub kind: T,
    #[serde(skip_serializing, skip_deserializing)]
    pub id: NodeId,
    pub span: Span,
}

impl<T> Node<T> {
    pub fn new(kind: T, span: Span) -> Self {
        let id = NodeId::create();
        Self { kind, id, span }
    }
}

impl<T> Spanned for Node<T> {
    fn span(&self) -> Span {
        self.span
    }
}

impl<T> From<&Node<T>> for Span {
    fn from(node: &Node<T>) -> Self {
        node.span
    }
}

impl<T> From<&Box<Node<T>>> for Span {
    fn from(node: &Box<Node<T>>) -> Self {
        node.span
    }
}

impl<T> From<&Node<T>> for NodeId {
    fn from(node: &Node<T>) -> Self {
        node.id
    }
}

impl<T> From<&Box<Node<T>>> for NodeId {
    fn from(node: &Box<Node<T>>) -> Self {
        node.id
    }
}
