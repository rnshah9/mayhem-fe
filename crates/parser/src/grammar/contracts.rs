use super::functions::parse_fn_def;
use super::types::{parse_event_def, parse_field, parse_opt_qualifier};

use crate::ast::{Contract, ContractStmt};
use crate::node::{Node, Span};
use crate::{ParseFailed, ParseResult, Parser, TokenKind};

// Rule: all "statement" level parse functions consume their trailing
// newline(s), either directly or via a function they call.
// This is required to parse an `if` block, because we need to peek past the
// trailing newlines to check whether it's followed by an `else` block, and is
// done for all statements for consistency.

/// Parse a contract definition.
/// # Panics
/// Panics if the next token isn't `contract`.
pub fn parse_contract_def(
    par: &mut Parser,
    contract_pub_qual: Option<Span>,
) -> ParseResult<Node<Contract>> {
    let contract_tok = par.assert(TokenKind::Contract);
    let contract_name = par.expect_with_notes(
        TokenKind::Name,
        "failed to parse contract definition",
        |_| vec!["Note: `contract` must be followed by a name, which must start with a letter and contain only letters, numbers, or underscores".into()],
    )?;

    let mut span = contract_tok.span + contract_name.span;
    par.enter_block(span, "contract definition")?;

    let mut fields = vec![];
    let mut defs = vec![];

    loop {
        par.eat_newlines();
        let mut pub_qual = parse_opt_qualifier(par, TokenKind::Pub);
        let const_qual = parse_opt_qualifier(par, TokenKind::Const);
        if pub_qual.is_none() && const_qual.is_some() && par.peek() == Some(TokenKind::Pub) {
            pub_qual = parse_opt_qualifier(par, TokenKind::Pub);
            par.error(
                pub_qual.unwrap() + const_qual,
                "`const pub` should be written `pub const`",
            );
        }

        match par.peek_or_err()? {
            TokenKind::Name => {
                let field = parse_field(par, pub_qual, const_qual)?;
                if !defs.is_empty() {
                    par.error(field.span, "contract field definitions must come before any function or event definitions");
                }
                fields.push(field);
            }
            TokenKind::Fn | TokenKind::Unsafe => {
                if let Some(span) = const_qual {
                    par.error(
                        span,
                        "`const` qualifier can't be used with function definitions",
                    );
                }
                defs.push(ContractStmt::Function(parse_fn_def(par, pub_qual)?));
            }
            TokenKind::Event => {
                if let Some(span) = pub_qual {
                    par.error(
                        span,
                        "`pub` qualifier can't be used with contract-level event definitions",
                    );
                }
                if let Some(span) = const_qual {
                    par.error(
                        span,
                        "`const` qualifier can't be used with event definitions",
                    );
                }
                defs.push(ContractStmt::Event(parse_event_def(par, None)?));
            }
            TokenKind::BraceClose => {
                span += par.next()?.span;
                break;
            }
            _ => {
                let tok = par.next()?;
                par.unexpected_token_error(
                    &tok,
                    "failed to parse contract definition body",
                    vec![],
                );
                return Err(ParseFailed);
            }
        };
    }

    Ok(Node::new(
        Contract {
            name: Node::new(contract_name.text.into(), contract_name.span),
            fields,
            body: defs,
            pub_qual: contract_pub_qual,
        },
        span,
    ))
}
