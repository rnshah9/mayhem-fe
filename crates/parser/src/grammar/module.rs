use super::contracts::parse_contract_def;
use super::expressions::parse_expr;
use super::functions::parse_fn_def;
use super::types::{
    parse_event_def, parse_impl_def, parse_path_tail, parse_struct_def, parse_trait_def,
    parse_type_alias, parse_type_desc,
};
use crate::ast::{ConstantDecl, Module, ModuleStmt, Pragma, Use, UseTree};
use crate::node::{Node, Span};
use crate::{Label, ParseFailed, ParseResult, Parser, TokenKind};

use semver::VersionReq;

/// Parse a [`Module`].
pub fn parse_module(par: &mut Parser) -> Node<Module> {
    let mut body = vec![];
    loop {
        match par.peek() {
            Some(TokenKind::Newline) => {
                par.next().unwrap();
            }
            None => break,
            Some(_) => {
                match parse_module_stmt(par) {
                    Ok(stmt) => body.push(stmt),
                    Err(_) => {
                        // TODO: capture a real span here
                        body.push(ModuleStmt::ParseError(Span::zero(par.file_id)));
                        break;
                    }
                };
            }
        }
    }
    let span = Span::zero(par.file_id) + body.first() + body.last();
    Node::new(Module { body }, span)
}

/// Parse a [`ModuleStmt`].
pub fn parse_module_stmt(par: &mut Parser) -> ParseResult<ModuleStmt> {
    let stmt = match par.peek_or_err()? {
        TokenKind::Pragma => ModuleStmt::Pragma(parse_pragma(par)?),
        TokenKind::Use => ModuleStmt::Use(parse_use(par)?),
        TokenKind::Contract => ModuleStmt::Contract(parse_contract_def(par, None)?),
        TokenKind::Struct => ModuleStmt::Struct(parse_struct_def(par, None)?),
        TokenKind::Trait => ModuleStmt::Trait(parse_trait_def(par, None)?),
        TokenKind::Impl => ModuleStmt::Impl(parse_impl_def(par)?),
        TokenKind::Type => ModuleStmt::TypeAlias(parse_type_alias(par, None)?),
        TokenKind::Const => ModuleStmt::Constant(parse_constant(par, None)?),

        TokenKind::Event => ModuleStmt::Event(parse_event_def(par, None)?),
        TokenKind::Pub => {
            let pub_span = par.next()?.span;
            match par.peek_or_err()? {
                TokenKind::Event => ModuleStmt::Event(parse_event_def(par, Some(pub_span))?),
                TokenKind::Fn | TokenKind::Unsafe => {
                    ModuleStmt::Function(parse_fn_def(par, Some(pub_span))?)
                }
                TokenKind::Struct => ModuleStmt::Struct(parse_struct_def(par, Some(pub_span))?),
                TokenKind::Trait => ModuleStmt::Trait(parse_trait_def(par, Some(pub_span))?),
                TokenKind::Type => ModuleStmt::TypeAlias(parse_type_alias(par, Some(pub_span))?),
                TokenKind::Const => ModuleStmt::Constant(parse_constant(par, Some(pub_span))?),
                TokenKind::Contract => {
                    ModuleStmt::Contract(parse_contract_def(par, Some(pub_span))?)
                }
                _ => {
                    let tok = par.next()?;
                    par.unexpected_token_error(
                        &tok,
                        "failed to parse module",
                        vec!["Note: expected `fn`".into()],
                    );
                    return Err(ParseFailed);
                }
            }
        }
        TokenKind::Fn | TokenKind::Unsafe => ModuleStmt::Function(parse_fn_def(par, None)?),
        _ => {
            let tok = par.next()?;
            par.unexpected_token_error(
                &tok,
                "failed to parse module",
                vec!["Note: expected import, contract, struct, type, const or event".into()],
            );
            return Err(ParseFailed);
        }
    };
    Ok(stmt)
}

/// Parse a constant, e.g. `const MAGIC_NUMBER: u256 = 4711`.
/// # Panics
/// Panics if the next token isn't `const`.
pub fn parse_constant(par: &mut Parser, pub_qual: Option<Span>) -> ParseResult<Node<ConstantDecl>> {
    let const_tok = par.assert(TokenKind::Const);
    let name = par.expect(TokenKind::Name, "failed to parse constant declaration")?;
    par.expect_with_notes(
        TokenKind::Colon,
        "failed to parse constant declaration",
        |_| {
            vec![
                "Note: constant name must be followed by a colon and a type description".into(),
                format!("Example: let `{}: u256 = 1000`", name.text),
            ]
        },
    )?;
    let typ = parse_type_desc(par)?;
    par.expect_with_notes(
        TokenKind::Eq,
        "failed to parse constant declaration",
        |_| {
            vec![
            "Note: the type of a constant must be followed by an equals sign and a value assignment"
                .into(),
                format!(
                    "Example: let `{}: u256 = 1000`",
                    name.text
                ),
        ]
        },
    )?;

    let exp = parse_expr(par)?;

    let span = const_tok.span + exp.span;
    Ok(Node::new(
        ConstantDecl {
            name: name.into(),
            typ,
            value: exp,
            pub_qual,
        },
        span,
    ))
}

/// Parse a `use` statement.
/// # Panics
/// Panics if the next token isn't `use`.
pub fn parse_use(par: &mut Parser) -> ParseResult<Node<Use>> {
    let use_tok = par.assert(TokenKind::Use);

    let tree = parse_use_tree(par)?;
    let tree_span = tree.span;

    Ok(Node::new(Use { tree }, use_tok.span + tree_span))
}

/// Parse a `use` tree.
pub fn parse_use_tree(par: &mut Parser) -> ParseResult<Node<UseTree>> {
    let (path, path_span, trailing_delim) = {
        let path_head =
            par.expect_with_notes(TokenKind::Name, "failed to parse `use` statement", |_| {
                vec![
                    "Note: `use` paths must start with a name".into(),
                    "Example: `use foo::bar`".into(),
                ]
            })?;
        parse_path_tail(par, path_head.into())
    };

    if trailing_delim.is_some() {
        match par.peek() {
            Some(TokenKind::BraceOpen) => {
                par.next()?;

                let mut children = vec![];
                let close_brace_span;

                loop {
                    children.push(parse_use_tree(par)?);
                    let tok = par.next()?;
                    match tok.kind {
                        TokenKind::Comma => {
                            continue;
                        }
                        TokenKind::BraceClose => {
                            close_brace_span = tok.span;
                            break;
                        }
                        _ => {
                            par.unexpected_token_error(
                                &tok,
                                "failed to parse `use` tree",
                                vec!["Note: expected a `,` or `}` token".to_string()],
                            );
                            return Err(ParseFailed);
                        }
                    }
                }

                Ok(Node::new(
                    UseTree::Nested {
                        prefix: path,
                        children,
                    },
                    close_brace_span,
                ))
            }
            Some(TokenKind::Star) => {
                par.next()?;
                Ok(Node::new(UseTree::Glob { prefix: path }, path_span))
            }
            _ => {
                let tok = par.next()?;
                par.unexpected_token_error(
                    &tok,
                    "failed to parse `use` tree",
                    vec!["Note: expected a `*`, `{` or name token".to_string()],
                );
                Err(ParseFailed)
            }
        }
    } else if par.peek() == Some(TokenKind::As) {
        par.next()?;

        let rename_tok = par.expect(TokenKind::Name, "failed to parse `use` tree")?;
        let span = path_span + rename_tok.span;
        let rename = Some(rename_tok.into());

        Ok(Node::new(UseTree::Simple { path, rename }, span))
    } else {
        Ok(Node::new(UseTree::Simple { path, rename: None }, path_span))
    }
}

/// Parse a `pragma <version-requirement>` statement.
pub fn parse_pragma(par: &mut Parser) -> ParseResult<Node<Pragma>> {
    let tok = par.assert(TokenKind::Pragma);
    assert_eq!(tok.text, "pragma");

    let mut version_string = String::new();
    let mut tokens = vec![];
    loop {
        match par.peek() {
            Some(TokenKind::Newline) => break,
            None => break,
            _ => {
                let tok = par.next()?;
                version_string.push_str(tok.text);
                tokens.push(tok);
            }
        }
    }

    let version_requirement_span = match (tokens.first(), tokens.last()) {
        (Some(first), Some(last)) => first.span + last.span,
        _ => {
            par.error(
                tok.span,
                "failed to parse pragma statement: missing version requirement",
            );
            return Err(ParseFailed);
        }
    };

    match VersionReq::parse(&version_string) {
        Ok(_) => Ok(Node::new(
            Pragma {
                version_requirement: Node::new(version_string.into(), version_requirement_span),
            },
            tok.span + version_requirement_span,
        )),
        Err(err) => {
            par.fancy_error(
                format!("failed to parse pragma statement: {}", err),
                vec![Label::primary(
                    version_requirement_span,
                    "Invalid version requirement",
                )],
                vec!["Example: `^0.5.0`".into()],
            );
            Err(ParseFailed)
        }
    }
}
