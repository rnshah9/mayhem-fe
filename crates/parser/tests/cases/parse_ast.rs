use fe_common::db::TestDb;
use fe_common::diagnostics::print_diagnostics;
use fe_common::utils::ron::to_ron_string_pretty;
use fe_common::SourceFileId;
use fe_parser::grammar::{expressions, functions, module, types};
use fe_parser::node::Node;
use fe_parser::{ast, ParseResult, Parser};
use insta::assert_snapshot;
use serde::Serialize;
use wasm_bindgen_test::wasm_bindgen_test;

pub fn ast_string<F, T>(test_name: &str, mut parse_fn: F, src: &str) -> String
where
    F: FnMut(&mut Parser) -> ParseResult<T>,
    T: Serialize,
{
    let mut db = TestDb::default();
    let id = SourceFileId::new_local(&mut db, test_name, src.into());
    let mut parser = Parser::new(id, src);

    if let Ok(ast) = parse_fn(&mut parser) {
        if !parser.diagnostics.is_empty() {
            print_diagnostics(&db, &parser.diagnostics);
            panic!("parse error");
        }
        to_ron_string_pretty(&ast).unwrap()
    } else {
        if parser.diagnostics.is_empty() {
            eprintln!("parsing failed, but no diagnostics were generated. this should be fixed.");
            let next = parser.next();
            if let Ok(tok) = next {
                parser.error(
                    tok.span,
                    "this is the next token at time of parsing failure",
                );
                print_diagnostics(&db, &parser.diagnostics);
            } else {
                eprintln!("parser is at end of file");
            }
        } else {
            print_diagnostics(&db, &parser.diagnostics);
        }
        panic!("parse failed");
    }
}

fn try_parse_module(par: &mut Parser) -> ParseResult<Node<ast::Module>> {
    // This is just to make `ast_string` above work; it's fine.
    Ok(module::parse_module(par))
}

macro_rules! test_parse {
    ($name:ident, $parse_fn:expr, $src:expr) => {
        #[test]
        #[wasm_bindgen_test]
        fn $name() {
            if cfg!(target_arch = "wasm32") {
                fe_common::assert_snapshot_wasm!(
                    concat!("snapshots/cases__parse_ast__", stringify!($name), ".snap"),
                    ast_string(stringify!($name), $parse_fn, $src)
                );
            } else {
                assert_snapshot!(ast_string(stringify!($name), $parse_fn, $src));
            }
        }
    };
}

test_parse! { expr_call1, expressions::parse_expr, "foo()" }
test_parse! { expr_call2, expressions::parse_expr, "foo(1,2,x:3)" }
test_parse! { expr_call3, expressions::parse_expr, "bing.foo<Bar>(x:3)" }
test_parse! { expr_call4, expressions::parse_expr, "bang.bing.foo<Bar, Baz>(26, 42)" }
test_parse! { expr_attr1, expressions::parse_expr, "foo.bar[0][y]" }
test_parse! { expr_attr2, expressions::parse_expr, "a[x].b[y](1)" }
test_parse! { expr_num1, expressions::parse_expr, "12345" }
test_parse! { expr_num2, expressions::parse_expr, "00001" }
test_parse! { expr_num3, expressions::parse_expr, "1000_000" }
test_parse! { expr_hex1, expressions::parse_expr, "0xbeefbeef" }
test_parse! { expr_hex2, expressions::parse_expr, "0xFEED1234" }
test_parse! { expr_path_call, expressions::parse_expr, "foo::bar::abc1()" }
test_parse! { expr_string, expressions::parse_expr, r#""hi \tmom\n""# }
test_parse! { expr_list, expressions::parse_expr, "[]" }
test_parse! { expr_list2, expressions::parse_expr, "[x, y, z,]" }
test_parse! { expr_repeat, expressions::parse_expr, "[true; 42]" }
test_parse! { expr_repeat2, expressions::parse_expr, "[5 + 4; 26]" }
test_parse! { expr_ternary, expressions::parse_expr, "x + 1 if y + 2 else z + 3" }
test_parse! { expr_group, expressions::parse_expr, "(1 + 2) * 3" }
test_parse! { expr_tuple1, expressions::parse_expr, "(1,)" }
test_parse! { expr_tuple2, expressions::parse_expr, "(1, 2, \n 3)" }
test_parse! { expr_tuple3, expressions::parse_expr, "(1, (2 + 3), (3 * 4, 5))" }
test_parse! { expr_unit, expressions::parse_expr, "()" }

test_parse! { ops_not, expressions::parse_expr, "x and not y" }
test_parse! { ops_math, expressions::parse_expr, "a + b * -c ** d / e % f" }
test_parse! { ops_neg, expressions::parse_expr, "-x" }
test_parse! { ops_bnot, expressions::parse_expr, "~x" }
// bitwise op precedence: shift > and > xor > or
test_parse! { ops_bit1, expressions::parse_expr, "a & b >> c" }
test_parse! { ops_bit2, expressions::parse_expr, "a ^ b & c" }
test_parse! { ops_bit3, expressions::parse_expr, "a | b ^ c" }
test_parse! { ops_shift, expressions::parse_expr, "a << b >> c" }
test_parse! { ops_bool, expressions::parse_expr, "a or b and c" }

test_parse! { stmt_assert_no_msg, functions::parse_stmt, "assert x == y" }
test_parse! { stmt_assert_msg, functions::parse_stmt, "assert x == y, z" }
test_parse! { stmt_aug_add, functions::parse_stmt, "x += y" }
test_parse! { stmt_aug_sub, functions::parse_stmt, "x -= y" }
test_parse! { stmt_aug_mul, functions::parse_stmt, "x *= y" }
test_parse! { stmt_aug_div, functions::parse_stmt, "x /= y" }
test_parse! { stmt_aug_mod, functions::parse_stmt, "x %= y" }
test_parse! { stmt_aug_and, functions::parse_stmt, "x &= y" }
test_parse! { stmt_aug_or, functions::parse_stmt, "x |= y" }
test_parse! { stmt_aug_xor, functions::parse_stmt, "x ^= y" }
test_parse! { stmt_aug_lsh, functions::parse_stmt, "x <<= y" }
test_parse! { stmt_aug_rsh, functions::parse_stmt, "x >>= y" }
test_parse! { stmt_aug_exp, functions::parse_stmt, "x **= y" }
test_parse! { stmt_emit1, functions::parse_stmt, "emit Foo()" }
test_parse! { stmt_emit2, functions::parse_stmt, "emit Foo(1, 2, x: y)" }
test_parse! { stmt_path_type, functions::parse_stmt, "let x: foo::Bar = foo::Bar(1, 2)" }
test_parse! { stmt_return1, functions::parse_stmt, "return" }
test_parse! { stmt_return2, functions::parse_stmt, "return x" }
test_parse! { stmt_return3, functions::parse_stmt, "return not x" }
test_parse! { stmt_revert1, functions::parse_stmt, "revert" }
test_parse! { stmt_revert2, functions::parse_stmt, "revert something" }

test_parse! { stmt_if, functions::parse_stmt, "if a { \n b }" }
test_parse! { stmt_if2, functions::parse_stmt, "if a { b } else if c { d } else if e { \n f } \n else {\n g }" }
test_parse! { stmt_while, functions::parse_stmt, "while a > 5 { \n a -= 1 }" }
test_parse! { stmt_for, functions::parse_stmt, "for a in b[0] {}" }
test_parse! { stmt_var_decl_name, functions::parse_stmt, "let foo: u256 = 1" }
test_parse! { stmt_var_decl_tuple, functions::parse_stmt, "let (foo, bar): (u256, u256) = (10, 10)" }
test_parse! { stmt_var_decl_tuples, functions::parse_stmt, "let (a, (b, (c, d))): x" }
test_parse! { stmt_semicolons, functions::parse_stmt, "if a { b; c; d; for x in y {}; }" }
test_parse! { type_def, try_parse_module, "type X = Map<address, u256>" }
test_parse! { pub_type_def, try_parse_module, "pub type X = Map<address, u256>" }
test_parse! { type_name, types::parse_type_desc, "MyType" }
test_parse! { type_array, types::parse_type_desc, "Array<address, 25>" }
test_parse! { type_3d, types::parse_type_desc, "Array<Array<Array<u256, 4>, 4>, 4>" }
test_parse! { type_string, types::parse_type_desc, "string<100>" }
test_parse! { type_generic, types::parse_type_desc, "foo<a, b<c>, Array<d, 10>>" }
test_parse! { type_generic_int, types::parse_type_desc, "foo<1, 2>" }
test_parse! { type_map1, types::parse_type_desc, "Map<address, u256>" }
test_parse! { type_map2, types::parse_type_desc, "Map<address, Map<u8, u256>>" }
test_parse! { type_map3, types::parse_type_desc, "Map<address, Map<u8, Map<u8, u8>>>" }
test_parse! { type_map4, types::parse_type_desc, "map < address , map < u8, u256 > >" }
test_parse! { type_tuple, types::parse_type_desc, "(u8, u16, address, Map<u8, u8>)" }
test_parse! { type_unit, types::parse_type_desc, "()" }

test_parse! { fn_def, try_parse_module, "fn transfer(from sender: address, to recip: address, _ val: u64) -> bool {\n false \n}"}

test_parse! { fn_def_generic, try_parse_module, "fn foo<T, R: Event>(this: T, that: R, _ val: u64) -> bool { false }"}
test_parse! { fn_def_pub, try_parse_module, "pub fn foo21(x: bool, y: address,) -> bool { x }"}
test_parse! { fn_def_unsafe, try_parse_module, "unsafe fn foo21(x: bool, y: address,) -> bool {\n x\n}"}
test_parse! { fn_def_pub_unsafe, try_parse_module, "pub unsafe fn foo21(x: bool, y: address,) -> bool{x}"}
test_parse! { event_def, try_parse_module, "event Foo {\n  x: address\n  idx y: u8\n}" }
test_parse! { empty_event_def, try_parse_module, "event Foo {}" }
test_parse! { pub_event_def, try_parse_module, "pub event Foo {\nx: address\nidx y: u8\n}" }
test_parse! { const_def, try_parse_module, "const FOO: i32 = 1" }
test_parse! { pub_const_def, try_parse_module, "pub const FOO: i32 = 1" }
test_parse! { pragma1, module::parse_pragma, "pragma 0.1.0" }
test_parse! { pragma2, module::parse_pragma, "pragma 0.1.0-alpha" }
test_parse! { pragma3, module::parse_pragma, "pragma >= 1.2, < 1.5" }

test_parse! { use_simple1, module::parse_use, "use foo::bar" }
test_parse! { use_simple2, module::parse_use, "use foo::bar as baz" }
test_parse! { use_glob, module::parse_use, "use foo::bar::*" }
test_parse! { use_nested1, module::parse_use, "use foo::bar::{bing::*, bang::big, bass as fish, bong::{hello as hi, goodbye}}" }
test_parse! { use_nested2, module::parse_use, r#"use std::bar::{
    bing::*,
    bad::{food as burger, barge::*, bill::bob::{jkl::*}},
    evm as mve
}"#
}
test_parse! { struct_def, try_parse_module, r#"struct S {
  x: address
  pub y: u8
  z: u8
  pub a: Map<u8, foo>

  pub fn foo(self) -> u8 {
    return self.z + self.y
  }
  unsafe fn bar() {}
}"# }
test_parse! { empty_struct_def, try_parse_module, "struct S {}" }

test_parse! { contract_def, try_parse_module, r#"contract Foo {
  x: address
  pub y: u8
  pub const z: Map<u8, address>

  pub fn foo() -> u8 {
    return 10
  }
  event Bar {
    idx from: address
  }
}
"# }

test_parse! { empty_contract_def, try_parse_module, "contract Foo {}" }

test_parse! { pub_contract_def, try_parse_module, r#"
pub contract Foo {
    pub fn foo() -> u8 {
      return 10
    }
}"# }

test_parse! { module_stmts, try_parse_module, r#"
pragma 0.5.0

use foo::bar::{
    bing as bong,
    food::*
}

type X = Map<u8, u16>

pub fn double(x: u8) -> u8 {
    return x * 2
}

fn secret() -> u8 { return 0xBEEF }

contract A {
    pub const x: u256 = 10
}

contract B {
    pub x: X
}"# }

test_parse! { guest_book, try_parse_module, r#"
type BookMsg = Array<bytes, 100>

contract GuestBook {
    pub guest_book: Map<address, BookMsg>

    event Signed {
        idx book_msg: BookMsg
    }

    pub fn sign(self, book_msg: BookMsg) {
        self.guest_book[msg.sender] = book_msg

        emit Signed(book_msg: book_msg)
    }
    pub fn get_msg(self, addr: address) -> BookMsg {
        return self.guest_book[addr]
    }
}"# }

test_parse! { module_level_events, try_parse_module, r#"
use std::context::Context

event Transfer {
    idx sender: address
    idx receiver: address
    value: u256
}
contract Foo {
    fn transfer(ctx: Context, to: address, value: u256) {
        emit Transfer(ctx, sender: msg.sender, receiver: to, value)
    }
}
"# }
