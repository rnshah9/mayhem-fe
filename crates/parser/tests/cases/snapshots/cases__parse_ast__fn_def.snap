---
source: crates/parser/tests/cases/parse_ast.rs
expression: "ast_string(stringify!(fn_def), try_parse_module,\n    \"fn transfer(from sender: address, to recip: address, _ val: u64) -> bool {\\n false \\n}\")"

---
Node(
  kind: Module(
    body: [
      Function(Node(
        kind: Function(
          sig: Node(
            kind: FunctionSignature(
              pub_: None,
              unsafe_: None,
              name: Node(
                kind: "transfer",
                span: Span(
                  start: 3,
                  end: 11,
                ),
              ),
              generic_params: Node(
                kind: [],
                span: Span(
                  start: 3,
                  end: 11,
                ),
              ),
              args: [
                Node(
                  kind: Regular(RegularFunctionArg(
                    label: Some(Node(
                      kind: "from",
                      span: Span(
                        start: 12,
                        end: 16,
                      ),
                    )),
                    name: Node(
                      kind: "sender",
                      span: Span(
                        start: 17,
                        end: 23,
                      ),
                    ),
                    typ: Node(
                      kind: Base(
                        base: "address",
                      ),
                      span: Span(
                        start: 25,
                        end: 32,
                      ),
                    ),
                  )),
                  span: Span(
                    start: 17,
                    end: 32,
                  ),
                ),
                Node(
                  kind: Regular(RegularFunctionArg(
                    label: Some(Node(
                      kind: "to",
                      span: Span(
                        start: 34,
                        end: 36,
                      ),
                    )),
                    name: Node(
                      kind: "recip",
                      span: Span(
                        start: 37,
                        end: 42,
                      ),
                    ),
                    typ: Node(
                      kind: Base(
                        base: "address",
                      ),
                      span: Span(
                        start: 44,
                        end: 51,
                      ),
                    ),
                  )),
                  span: Span(
                    start: 37,
                    end: 51,
                  ),
                ),
                Node(
                  kind: Regular(RegularFunctionArg(
                    label: Some(Node(
                      kind: "_",
                      span: Span(
                        start: 53,
                        end: 54,
                      ),
                    )),
                    name: Node(
                      kind: "val",
                      span: Span(
                        start: 55,
                        end: 58,
                      ),
                    ),
                    typ: Node(
                      kind: Base(
                        base: "u64",
                      ),
                      span: Span(
                        start: 60,
                        end: 63,
                      ),
                    ),
                  )),
                  span: Span(
                    start: 55,
                    end: 63,
                  ),
                ),
              ],
              return_type: Some(Node(
                kind: Base(
                  base: "bool",
                ),
                span: Span(
                  start: 68,
                  end: 72,
                ),
              )),
            ),
            span: Span(
              start: 0,
              end: 72,
            ),
          ),
          body: [
            Node(
              kind: Expr(
                value: Node(
                  kind: Bool(false),
                  span: Span(
                    start: 76,
                    end: 81,
                  ),
                ),
              ),
              span: Span(
                start: 76,
                end: 81,
              ),
            ),
          ],
        ),
        span: Span(
          start: 0,
          end: 84,
        ),
      )),
    ],
  ),
  span: Span(
    start: 0,
    end: 84,
  ),
)
